operator/metadata/updater.go

```diff
@@ -0,0 +1,260 @@
+		zap.Int("count", len(metadata)),
+		zap.Int("shares", len(pubKeys)),
+	)
+
+	updateStart := time.Now()
+	// TODO: Refactor share storage to support passing context.
+	if err := u.shareStorage.UpdateValidatorsMetadata(metadata); err != nil {
+		u.logger.Error("failed to update validators metadata after setup",
+			zap.Int("shares", len(pubKeys)),
+			fields.Took(time.Since(updateStart)),
+			zap.Error(err),
+		)
+		return metadata, fmt.Errorf("update metadata: %w", err)
+	}
+
+	u.logger.Debug("🆕 updated validators metadata in storage",
+		fields.Took(time.Since(updateStart)),
+		zap.Int("count", len(metadata)),
+		zap.Int("shares", len(pubKeys)),
+	)
+	return metadata, nil
+}
+
+type Update struct {
+	IndicesBefore []phase0.ValidatorIndex
+	IndicesAfter  []phase0.ValidatorIndex
+	Validators    Validators
+}
+
+func (u *Updater) Stream(ctx context.Context) <-chan Update {
+	metadataUpdates := make(chan Update, streamChanSize)
+
+	go func() {
+		defer close(metadataUpdates)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				shares, err := u.sendUpdate(ctx, metadataUpdates)
+				if err != nil {
+					u.logger.Warn("failed to update validators metadata",
+						zap.Int("shares", len(shares)),
+						zap.Error(err),
+					)
+					continue
+				}
+
+				// Only sleep if there aren't more validators to fetch metadata for.
+				if len(shares) < batchSize {
+					time.Sleep(u.streamInterval)
+				}
+			}
+		}
+	}()
+
+	return metadataUpdates
+}
+
+func (u *Updater) sendUpdate(ctx context.Context, updates chan<- Update) ([]*ssvtypes.SSVShare, error) {
+	shares := u.sharesForUpdate()
+	if len(shares) == 0 {
+		return shares, nil
+	}
+
+	pubKeys := make([]spectypes.ValidatorPK, len(shares))
+	for i, s := range shares {
+		pubKeys[i] = s.ValidatorPubKey
+	}
+
+	indicesBefore := u.allActiveIndices(u.beaconNetwork.GetBeaconNetwork().EstimatedCurrentEpoch())
+
+	updatedMetadata, err := u.Update(ctx, pubKeys)
+	if err != nil {
+		return shares, fmt.Errorf("update metadata: %w", err)
+	}
+
+	indicesAfter := u.allActiveIndices(u.beaconNetwork.GetBeaconNetwork().EstimatedCurrentEpoch())
+
+	update := Update{
+		IndicesBefore: indicesBefore,
+		IndicesAfter:  indicesAfter,
+		Validators:    updatedMetadata,
+	}
+
+	timer := time.NewTimer(u.updateSendTimeout)
+	defer timer.Stop()
+
+	select {
+	case <-ctx.Done():
+		return shares, ctx.Err()
+	case <-timer.C:
+		return shares, fmt.Errorf("timed out waiting for sending update")
+	case updates <- update:
+		return shares, nil
+	}
+}
+
+func (u *Updater) sharesForUpdate() []*ssvtypes.SSVShare {
+	var existingShares, newShares []*ssvtypes.SSVShare
+	u.shareStorage.Range(nil, func(share *ssvtypes.SSVShare) bool {
+		if share.Liquidated {
+			return true
+		}
+		if share.BeaconMetadata == nil && share.MetadataLastUpdated().IsZero() {
+			newShares = append(newShares, share)
+		} else if time.Since(share.MetadataLastUpdated()) > u.updateInterval {
+			existingShares = append(existingShares, share)
```

Comment:
`staleShares` is fine, but I'm not sure I agree with the latter because shares become stale after the update interval, and if we rename it, it won't be readable in other places

Reply:
Maybe `shareMetadataStaleAfter` is not the best name (just an example),

what I'm saying is `updateInterval` is too generic of a name that could mean many things - hence would be nice to make it more concrete. What it really represents is "the duration (span of time) after which share metadata is considered stale" - `shareMetadataStillFresh` / `shareMetadataStillFreshDuration`

this one is pretty hard to describe with just name though :)